Incorporating "matching.id" - Phase I

AaronPotvien opened the pull request:
- Took on a "do no harm" approach to incorporating a `matching.id` variable, so that if `matching.id` is unspecified (`NULL`), then all behavior *should* be exactly as before.

- The update should achieve incorporating the matching ID in the folds of all `cv.glmnet` calls (i.e. matched groups are always members of the same fold).

- This also happens in the bootstrapped sampling routines, where it's slightly more complicated: If a matching ID is supplied, then `train.fraction` is interpreted as the fraction of clusters instead of observations. Additionally, when drawn, clusters are assigned a new `matching.id` to reflect their draws as "new" random samples in the call to be passed back to `fit.subgroup`.

- There are two unfinished aspects to this work:

1) Currently, setting `propensity.func()` with or without `matching.id` has no practical consequence. However, the conditional logic is in place so that different behavior can be incorporated when a user specifies any combination of `matching.id` (Y/N) x `propensity.func` (Y/N) x `n.trt` (==2 or more). Before deciding on the logic for every possibility, I thought it would be best to first reach this checkpoint.

2) It's unclear how to connect the `matching.id` to the `gbm` function in the same way that it was done for `cv.glmnet` to assign each cluster to its own fold. While this function can take a  `cv.folds` argument, it doesn't seem to allow for direct interface with the folds themselves. This is in contrast to `cv.glmnet`, which does allow for such specification via `foldid`. If this can't be resolved, it may be worth warning the user that the cross-validation technique in `gbm.perf` constructs folds without regard to the matched data.

AaronPotvien committed:
fit_subgroup to allow for matching.id argument

As a first step, simply defines propensity.func() as a function of 2 or 3 arguments, depending on whether matching.id is specified. Logic is needed about how to handle each Yes/No combination: (n.trt == 2 or more) x (propensity.func specified) x (matching.id specified)

Some basic validity checking for matching.id argument added also.

AaronPotvien committed:
Update fit_losses to account for matching.id

Logic added in to have clusters belong to the same folds whenever LASSO is used so that every matched group member always belongs to the same fold. If user specifies both a matching.id and a foldid, the user-specified foldid is currently ignored.

Still need to consider if/how the matching ID should play a role in the cross-validation piece of the GBM models.

AaronPotvien committed:
Update fit_subgroup.R

AaronPotvien committed:
Update fit_subgroup.R

AaronPotvien committed:
Update fit_losses.R

AaronPotvien committed:
Added matching.id logic to validate.subgroup

If a matching ID is detected, sampling is done at the cluster level as defined by this ID.

Matching IDs are reassigned based on the draw so that even if identical clusters as sampled, they are treated as separate clusters from a random sample perspective.

AaronPotvien committed:
Update fit_subgroup.R

AaronPotvien committed:
Update fit_losses.R

AaronPotvien committed:
Update validate_subgroup.R

AaronPotvien committed:
Update fit_subgroup.R

jaredhuling commented:
It seems like there is some error in how "nfolds" for cv.glmnet() is handled. somewhere along the way it seems like it's getting passed to cv.glmnet() twice

I like the way this is handled. I think just giving a warning wrt gbm is fine in this case, as it seems like there is nothing that can be done about it easily

AaronPotvien committed:
Fixed bug in sq_loss_lasso_gam that caused nfolds to be set twice.

Changed order of operations in fit_sq_loss_lasso_gam to fix bug. Also removed checking whether options are supplied via list.dots, as nfolds and foldid are now always supplied via list.dots.

AaronPotvien committed:
Added warning message for GBM Models 

Added the following warning message if the user tries to input a matching.id while using a GBM-based loss:

"Matched groups are not guaranteed to remain matched in the cross-validation procedure using GBM models."
